### PR TITLE
refactor(relay): introduce type-safe `Server` APIs

### PR DESCRIPTION
**Pull request by thomaseizinger**
We introduce dedicated types for each message that the `Server` can handle. This allows us to make the functions public because the type-system now guarantees that those are either parsed from bytes or constructed with the correct data.

The latter will be useful to write tests against a richer API.

### CODE REVIEW

File: rust/relay/tests/regression.rs [hunk 1]
```diff
@@ -123,11 +125,6 @@ fn ping_pong_relay() {
             Output::send_message("127.0.0.1:42677","010800002112a442dc5c115f6b727e25a54b55d3")
         ]),
         (
-        Input::client("127.0.0.1:42677","001600242112a4421b6fb3ce9cefcd57ef3a9edb00130009484f4c4550554e4348000000001200080001f2405e12a443802800041bfe0967", now),
-        &[
-            Output::send_message("127.0.0.1:42677","011600142112a4421b6fb3ce9cefcd57ef3a9edb0009000f00000400426164205265717565737400")
-        ]),
-        (
```
Comment (thomaseizinger) [inline rust/relay/tests/regression.rs:130]: This is deleted because it is a `SEND` indication which we don't support.

**Comment by thomaseizinger** [comment]
Friendly ping @firezone/engineering, this is ready for review :)

**Review by conectado — APPROVED** [review]
LGTM

**Comment by thomaseizinger** [comment]
Not sure how I can fix the failing CI?

**Pull request merged**